Add Mesos hyperkube minion server

sttts opened the pull request:
The minion hyperkube server will launch the executor and the proxy and rotate their logs, as a replacement of a S6 setup.

sttts committed:
Add scheduler flag --executor-proxy-logv

sttts committed:
Share KM_* constants

sttts committed:
Add lumberjack to Godeps

sttts committed:
Add Mesos hyperkube minion server

The minion server will
- launch the proxy and executor
- relaunch them when they terminate uncleanly
- logrotate their logs.

It is a replacement for a full-blown init process like s6 which is not necessary
in this case.

sttts commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -0,0 +1,223 @@
+		log.Infof("wrote %d bytes to %v", written, logFileName)
+	}()
+
+	// if the proxy fails to start then we exit the executor, otherwise
```
This error handling must be updated to the new architecture.

jdef commented:
integration test failing in shippable for one of the two builds (so, unrelated?):
```
I0728 20:31:47.262371 32362 config.go:253] Setting pods for source file : {[] 0 file}
I0728 20:31:48.262781 32362 config.go:253] Setting pods for source file : {[] 0 file}
F0728 20:31:48.575741 32362 integration.go:879] FAILED: (Scheduler probably didn't process deletion of 'phantom.bar') Pod never started running: timed out waiting for the condition
!!! Error in ./hack/test-integration.sh:51
```

jdef commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -0,0 +1,223 @@
+			}
+		}()
+
+		writer := lumberjack.Logger{
```
can we make `writer` a pointer instead?

jdef commented on contrib/mesos/pkg/scheduler/service/service.go:
```diff
@@ -264,17 +266,11 @@ func (s *SchedulerServer) prepareExecutorInfo(hks hyperkube.Interface) (*mesos.E
 		ci.Uris = append(ci.Uris, &mesos.CommandInfo_URI{Value: proto.String(uri), Executable: proto.Bool(true)})
 		ci.Value = proto.String(fmt.Sprintf("./%s", executorCmd))
-	} else if !hks.FindServer(KM_EXECUTOR) {
+	} else if !hks.FindServer(hyperkube.KM_EXECUTOR) {
```
should probably check for KM_MINION here since that's what will be called

jdef commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -0,0 +1,223 @@
+	// create command
+	cmd := exec.Command(ms.kmBinary, kmArgs...)
+	if _, err := cmd.StdoutPipe(); err != nil {
+		log.Fatal(err)
```
can we just return `err` instead?

jdef commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -0,0 +1,223 @@
+	}
+	stderrLogs, err := cmd.StderrPipe()
+	if err != nil {
+		log.Fatal(err)
```
return `err`?

jdef commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -0,0 +1,223 @@
+
+		writer := lumberjack.Logger{
+			Filename:   logFileName,
+			MaxSize:    1,
```
magic numbers (1, 3). extract constants (as defaults) and allow configuration via params.

jdef commented:
reviewed, a few minor nits. otherwise lgtm. have not tested

sttts committed:
Disable private mount ns for now in Mesos hyperkube minion server

Until Docker learns parent mount namespace customization the container will
always have the root ns as a parent, not the one of the km minion. Hence, the
kubelet (which lives in the km minion mount ns) will create mounts that cannot
be seen by the Docker containers.

This feature can be enabled again when Docker learns to explicitly set the
parent mount ns, in analogy to the parent cgroup.

sttts commented:
@jdef PTAL

The command line handling in the scheduler is not really elegant right now. This really asks for a multiflags package (not in the limits of the PR) which can handle command lines like
```
km scheduler --logv=2 executor --logv=3 proxy --logv=9
```
with scheduler as the primary command which somehow launches the executor and proxy.

jdef commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -0,0 +1,269 @@
+
+	// wait until minion exit is requested
+	err = <-ms.exit
+	close(ms.exit)
```
possible data race: closing a chan could happen at the same time as a chan write. plus, it's not really needed here: we already have .done for listeners that want to wait for Run() to complete. suggest removing this line.

sttts replied:
A write to a closed channel will either block (and then gets unblocked by the close) or fails because the channel is closed. No? Which race-condition do you have in mind?

jdef replied:
the race detector has sounded the alarm many times on me for this.
solutions are:
(a) only a single channel writer should exist and it closes the chan
(b) if there are multiple writers, then synchronize around write ops and
close
(c) don't close the chan


On Wed, Jul 29, 2015 at 5:13 PM, Dr. Stefan Schimanski <
notifications@github.com> wrote:

> In contrib/mesos/pkg/minion/server.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11928#discussion_r35812598>
> :
>
> > +	if err != nil {
> > +		// required for k8sm since we need to send api.Binding information
> > +		// back to the apiserver
> > +		log.Fatalf("No API client: %v", err)
> > +	}
> > +	ms.clientConfig = clientConfig
> > +
> > +	// run subprocess until
> > +	if ms.runProxy {
> > +		go runtime.Until(ms.launchProxyServer, 5*time.Second, ms.done)
> > +	}
> > +	go runtime.Until(ms.launchExecutorServer, 5*time.Second, ms.done)
> > +
> > +	// wait until minion exit is requested
> > +	err = <-ms.exit
> > +	close(ms.exit)
>
> A write to a closed channel will either block (and then gets unblocked by
> the close) or fails because the channel is closed. No? Which race-condition
> do you have in mind?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11928/files#r35812598>
> .
>

sttts replied:
Then I learned something new about channels. Thanks.

sttts replied:
Axiom 3 of http://dave.cheney.net/2014/03/19/channel-axioms

jdef commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -0,0 +1,269 @@
+	// wait for the proxy process to end (and release resources after).
+	if err := cmd.Start(); err != nil {
+		// fatal error => terminate minion
+		ms.exit <- err
```
ms.exit is only read from once. why write to it twice?

sttts replied:
Good spot. A typo...

jdef commented:
few minor comments. otherwise lgtm.

sttts commented:
Addressed all remaining minor nits. Rebased and squashed (as far as it makes sense).

@jdef or @karlkfi Ready to be merged. Can you assign to @davidopp?

This PR only modified contrib/mesos files with the exception of the new lumberjack.v2 logging library as godep.

tsenart commented on contrib/mesos/pkg/hyperkube/hyperkube.go:
```diff
@@ -0,0 +1,26 @@
+	KM_EXECUTOR           = "executor"
+	KM_MINION             = "minion"
+	KM_PROXY              = "proxy"
+	KM_SCHEDULER          = "scheduler"
```
http://golang.org/doc/effective_go.html#mixed-caps

tsenart commented on contrib/mesos/pkg/minion/config/config.go:
```diff
@@ -0,0 +1,30 @@
+	DefaultLogMaxAgeInDays = 7
+)
+
+func DefaultLogMaxSize() resource.Quantity {
```
Can we comment exported types please?

tsenart commented on contrib/mesos/pkg/minion/mountns_darwin.go:
```diff
@@ -0,0 +1,25 @@
+	log "github.com/golang/glog"
+)
+
+func enterPrivateMountNamespace() {
```
Wouldn't it make sense to have a `mountns.go` that is compiled on all OSes besides Linux instead of just Mac OSX?

sttts replied:
mountns.go will also be compiled for linux, leading to `enterPrivateMountNamespace redeclared in this block`. Is there a way to define a fallback if the right `mountns_<os>.go` does not exist?

jdef replied:
typically via something like mountns_unsupported.go, which defines build flags -linux -darwin

tsenart commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -0,0 +1,268 @@
+)
+
+type MinionServer struct {
+	KubeletExecutorServer *exservice.KubeletExecutorServer
```
Why not embed?

sttts replied:
I prefer that to be explicit because it's not really clean right now to do this instantiation of `KubeletExecutorServer` at all. With a multiflags approach for the km tool, this will go away.

tsenart commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -0,0 +1,268 @@
+	proxyBindall bool
+}
+
+func NewMinionServer() *MinionServer {
```
Comments please.

tsenart commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -0,0 +1,268 @@
+	}
+
+	log.Info("Executor exited cleanly, stopping the minion")
+	ms.exit <- nil
```
`close(ms.exit)`?

sttts replied:
added a comment. Closing ms.exit leads to panics for the go routines writing to it. We had that before and changed it.

tsenart commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -0,0 +1,268 @@
+
+// runs the main kubelet loop, closing the kubeletFinished chan when the loop exits.
+// never returns.
+func (ms *MinionServer) Run(hks hyperkube.Interface, _ []string) error {
```
`defer close(ms.done)`

sttts replied:
done

tsenart commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -0,0 +1,268 @@
+	// wait until minion exit is requested
+	err = <-ms.exit
+	close(ms.done)
+	return err
```
`return <-ms.exit`

tsenart commented on contrib/mesos/pkg/scheduler/service/service.go:
```diff
@@ -292,18 +307,16 @@ func (s *SchedulerServer) prepareExecutorInfo(hks hyperkube.Interface) (*mesos.E
+		ci.Arguments = append(ci.Arguments, fmt.Sprintf("--proxy-bindall=%v", s.ProxyBindall))
+		ci.Arguments = append(ci.Arguments, fmt.Sprintf("--proxy-logv=%d", s.ProxyLogV))
+
+		ci.Arguments = append(ci.Arguments, fmt.Sprintf("--max-log-size=%v", s.MinionLogMaxSize.String()))
```
Is the `.String()` call necessary here?

sttts replied:
No, it's implicit by the %v semantics.

tsenart commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -0,0 +1,268 @@
+	executorArgs, _ := filterArgsByFlagSet(allArgs, executorFlags)
+
+	// run executor and quit minion server when this exits cleanly
+	err := ms.launchHyperkubeServer(hyperkube.KM_EXECUTOR, &executorArgs, "executor.log")
```
I suppose we could parameterize the log file name.

sttts replied:
Was thinking about that, but decided against it. People will start to define the same file for multiple processes. Unnecessary complexity for hardly a win.

tsenart commented on contrib/mesos/pkg/minion/server.go:
```diff
@@ -0,0 +1,268 @@
+	allArgs := os.Args[1:]
+
+	// filter out minion flags, leaving those for the executor
+	executorFlags := pflag.NewFlagSet("executor flags", pflag.ContinueOnError)
```
Flagset names usually refer to the command they belong to.`s/executor flags/executor/` ?

sttts commented:
Rebased to master and addressed @tsenart's comment (if not state otherwise above).

Ready for merge.

mikedanese commented on cmd/hyperkube/kube-proxy.go:
```diff
@@ -20,6 +20,7 @@ package main
 
 import (
 	kubeproxy "github.com/GoogleCloudPlatform/kubernetes/cmd/kube-proxy/app"
+	"github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/hyperkube"
```
These dependencies are no good. We need a better spot to put these constants so we don't link in mesoskube to hyperkube.

sttts replied:
yeah. They shouldn't be there for sure. Fixing this. Sorry for the inconvenience.

mikedanese replied:
no worries.

mikedanese commented on cmd/hyperkube/kube-scheduler.go:
```diff
@@ -19,6 +19,7 @@ limitations under the License.
 package main
 
 import (
+	"github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/hyperkube"
```
same as comment above

sttts replied:
same as above.

mikedanese commented:
This broke 'make quick-release'
